src/java.base/share/classes/javax/security/auth/Subject.java

```diff
@@ -319,6 +329,181 @@ public Subject run() {
+     * @throws NullPointerException if the {@code Callable} is {@code null}.
+     * @throws CompletionException if {@action.call()} throws a checked
+     *          exception, which will be the cause of this
+     *          {@code CompletionException}.
+     * @see #current()
+     * @since 18
+     */
+    public static <T> T callAs(final Subject subject,
+            final Callable<T> action) throws CompletionException {
+        Objects.requireNonNull(action);
+        return switch (SUBJECT_STORAGE) {
+            case ACC -> {
+                try {
+                    PrivilegedExceptionAction<T> pa = () -> action.call();
+                    yield legacyDoAs(subject, pa);
+                } catch (PrivilegedActionException e) {
+                    throw new CompletionException(e.getException());
+                }
+            }
+            case SCOPE_LOCAL -> {
+                yield null;
+//                try {
+//                    yield ScopeLocal.where(SUBJECT_SCOPE, subject)
+//                            .call(action::call);
+//                } catch (RuntimeException re) {
+//                    throw re;
+//                } catch (Exception e) {
+//                    throw new CompletionException(e);
+//                }
+            }
+            case THREAD_LOCAL -> {
+                Subject oldSubject = SUBJECT_THREAD_LOCAL.get();
+                SUBJECT_THREAD_LOCAL.set(subject);
+                try {
+                    yield action.call();
+                } catch (RuntimeException re) {
+                    throw re;
+                } catch (Exception e) {
+                    throw new CompletionException(e);
+                } finally {
+                    SUBJECT_THREAD_LOCAL.set(oldSubject);
+                }
+            }
+        };
+    }
+
+    /**
+     * Execute a {@code Supplier} with {@code subject} as the
+     * current installed subject.
+     *
+     * @implNote
+     * In this implementation, if a {@code SecurityManager} is allowed,
+     * this method is implemented based on
+     * {@link #doAs(Subject, PrivilegedAction)}. Otherwise, it
+     * installs {@code subject} into an inheritable {@code ThreadLocal}
+     * before running {@code action} and restore the previous value of
+     * the {@code ThreadLocal} upon return.
+     *
+     * @param subject the intended current installed subject for {@code action}.
+     *                Can be {@code null}.
+     * @param action the code to be run with {@code subject} as its current
+     *               installed subject. Must not be {@code null}.
+     * @param <T> the type of value returned by the {@code get} method
+     *            of {@code action}
+     * @return the value returned by the {@code get} method of {@code action}
+     * @throws NullPointerException if the {@code Supplier} is {@code null}.
+     * @see #current()
+     * @since 18
+     */
+    public static <T> T getAs(final Subject subject,
```

Comment:
Would prefer if we just have one `callAs` method. It might be less of an issue now if we don't need to support the SM case.